rules/cakephp/src/Rector/MethodCall/ArrayToFluentCallRector.php

```diff
@@ -0,0 +1,223 @@
+
+    /**
+     * @param mixed[] $factoryMethods
+     */
+    public function __construct(array $configurableClasses = [], array $factoryMethods = [])
+    {
+        $this->configurableClasses = $configurableClasses;
+        $this->factoryMethods = $factoryMethods;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition('Moves array options to fluent setter method calls.', [
+            new CodeSample(
+                <<<'PHP'
+class ArticlesTable extends \Cake\ORM\Table
+{
+    public function initialize(array $config)
+    {
+        $this->belongsTo('Authors', [
+            'foreignKey' => 'author_id',
+            'propertyName' => 'person'
+        ]);
+    }
+}
+PHP
+                ,
+                <<<'PHP'
+class ArticlesTable extends \Cake\ORM\Table
+{
+    public function initialize(array $config)
+    {
+        $this->belongsTo('Authors')
+            ->setForeignKey('author_id')
+            ->setProperty('person');
+    }
+}
+PHP
+            ),
+        ]);
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [MethodCall::class];
+    }
+
+    /**
+     * @param MethodCall $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $config = $this->matchTypeAndMethodName($node);
+        if ($config === null) {
+            return null;
+        }
+
+        [$argument, $class] = $config;
+
+        $fluentMethods = $this->resolveFluentMethodCalls($class);
+        if (! $fluentMethods) {
+            return null;
+        }
+
+        return $this->replaceArrayToFluentMethodCalls($node, $argument, $fluentMethods);
+    }
+
+    /**
+     * @return mixed[]|null
+     */
+    private function matchTypeAndMethodName(MethodCall $methodCall): ?array
+    {
+        foreach ($this->factoryMethods as $className => $methodName) {
+            if (! $this->isObjectType($methodCall->var, $className)) {
+                continue;
+            }
+
+            /** @var string[] $methodNames */
+            $methodNames = array_keys($methodName);
+            if (! $this->isNames($methodCall->name, $methodNames)) {
+                continue;
+            }
+
+            $currentMethodName = $this->getName($methodCall->name);
+            if ($currentMethodName === null) {
+                continue;
+            }
+
+            $config = $methodName[$currentMethodName];
+            $argument = $config[self::ARGUMENT] ?? 1;
+            $class = $config[self::CLASSNAME];
+
+            return [$argument, $class];
+        }
+
+        return null;
+    }
+
+    private function resolveFluentMethodCalls(string $class)
+    {
+        if (! isset($this->configurableClasses[$class])) {
+            return [];
+        }
+
+        $map = $this->configurableClasses[$class];
+
+        if (isset($map[self::EXTEND])) {
+            $parent = $map[self::EXTEND];
+            unset($map[self::EXTEND]);
```

Comment:
Some classes inherits parent class, and also array options.

To avoid duplicating key-setter map from, maps can inherit/extend map from another class.

Eg. `HasMany` inherits all options from `Association` and adds few more own options:

```yml
            'Cake\ORM\Association\HasMany':
                _extend: 'Cake\ORM\Association'
                saveStrategy: setSaveStrategy
                sort: setSort 
```

Reply:
Ah, that's unclear from the config. Also, some methods can be called __extend, _extend etc.

Configs should be as dummy as possible, without any magic. Please keep this duplicated, to avoid possible bugs.

Reply:
Btw, isn't `Cake\ORM\Association\HasMany` child class of `Cake\ORM\Association`?

If so, there is no need to duplicate it, as the type is checked, not the class string.
Try it in tests.

Reply:
I removed `_extend` at all.

While some options are valid for `BelongsToMany`/`HasMany` only, I merged all associations to one list. 

If invalid option like `sort` will be used on `BelongsTo`/`HasOne`, Rector will still generate non-existing method call `setSort()`, but IDE/static analyzer will report that and help find those previously ignored array options.

Reply:
> If invalid option like sort will be used on BelongsTo/HasOne, Rector will still generate non-existing method call setSort(), but IDE/static analyzer will report that and help find those previously ignored array options.

Are you ok with this approach? If so, I can merge it

Reply:
Yes.